### PR TITLE
SI-2991 No Java tupling

### PR DESCRIPTION
**Pull request by som-snytt**
SI-2991 No Java tupling

A compiler option `-Yno-java-tupling` ensures that
tupling is not used to make a Java method applicable.

This includes applicability tests during overload
resolution, so that Java API that was previously
ambiguous can be invoked from Scala.

For example, Java will disambiguate `f(A)` and `f(A, A*)` 
even though ambiguous in Scala. Turning off Java tupling
allows invoking the single-parameter method defined in Java.

Anticipating deprecation of the feature, which was
never specked, `-Yno-tupling` turns it off entirely.

Currently, `-Yno-adapted-args` invalidates an
adaptation only after overload resolution.

Tupling becomes a Context mode which defaults off.

Review by @retronym

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -178,6 +178,17 @@ trait Infer extends Checkable {
     def context: Context
     import InferErrorGen._
 
+    // flag used to turn off tupling when alternatives of an overload include a Java-defined method
+    private[this] var tupleForApplicability = true
+    def withoutTuplingForApplicability[A](body: => A): A =
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: This seems like a new place to store this sort of state. Usually I would expect it to be a `ContextMode` or a `Mode` bit. I'm not sure that storing this here is necessarily incorrect, but I can't really be sure.

The class of bug that worries me is when, in the apparent seclusion of `try body finally ...`, a type completer is triggered that bears witness the state you're toggling.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 2]
```diff
@@ -1396,13 +1407,25 @@ trait Infer extends Checkable {
...
+
         private def followType(sym: Symbol) = followApply(pre memberType sym)
         // separate method to help the inliner
         private def isAltApplicable(pt: Type)(alt: Symbol) = context inSilentMode { isApplicable(undetparams, followType(alt), argtpes, pt) && !context.reporter.hasErrors }
         private def rankAlternatives(sym1: Symbol, sym2: Symbol) = isStrictlyMoreSpecific(followType(sym1), followType(sym2), sym1, sym2)
         private def bestForExpectedType(pt: Type, isLastTry: Boolean): Unit = {
           val applicable  = overloadsToConsiderBySpecificity(alts filter isAltApplicable(pt), argtpes, varargsStar)
-          val ranked      = bestAlternatives(applicable)(rankAlternatives)
+          val infected    = applicable exists (_.isJavaDefined)
+          val ranked      = atRetryDepth(infected) { bestAlternatives(applicable)(rankAlternatives) }
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: I'd prefer if any new code was quarantined, performance wise, along the lines of:

``` scala
def findBest = bestAlternatives(applicable)(rankAlternatives)

if (settings.Y...) .... findBest ..
   else findBest
```

As it stands, this allocates a closure (for by-name param `body`), an `ObjectRef` (for `var retries`) before it checks the setting.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 3]
```diff
@@ -1396,13 +1407,25 @@ trait Infer extends Checkable {
...
+        // if at least one retry is required, and the overload includes a Java-defined alternative
+        // proceed without tupling for applicability (if flag is enabled)
+        private[this] var retries = 0
+        private def atRetryDepth[A](cond: Boolean)(body: => A): A = {
+          val res =
+            if (cond && retries > 0) withoutTuplingForApplicability(body)
+            else body
+          retries += 1
+          res
+        }
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: I don't really follow this mechanism. What is being retried? Is this related to `isLastTry`?

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 1 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: In fact, I was prompted to look at this PR this afternoon, and didn't like the vars. I always get an "I wrote that?" feeling, but this wasn't pleasant. I couldn't see init order and usages at my accustomed glance; maybe I can find the guy who wrote it and get him to work it up a bit.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 2 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Yes.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 3 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: I can't get ahold of the guy. He might be either out of the country or deep in-country. It's too bad he didn't leave a few more tests, as clues to what he might have been thinking.

It looks like you get one chance for your Java method to be chosen: implicits are off and `pt` is what it is. Thereafter, you don't want expected type messing with overload resolution involving a Java-defined method, perhaps the way you don't want an ump or ref's call deciding a sports match in the final moments. I'll see if there's a test case.

**Comment by som-snytt** [comment]
To add tests, restructure.

**Comment by som-snytt** [comment]
Not sure if this is quite there yet, needs more tests at least, but thx @retronym for pointing out how bad the first cut was. Some external contributions, albeit appreciated, sneak under the bar.

**Comment by adriaanm** [comment]
Thanks for digging in, here (though not so as to fit under the bar)! It would be nice to at least have some way of calling those enterprisey methods.

**Comment by som-snytt** [comment]
@adriaanm Pls pretend I don't remember anything older than a day. Because I don't. I almost left my baby on the bus when she was a day old. So which methods do you mean?

**Comment by som-snytt** [comment]
Oh, duh, I just got it. Sorry, I'm back home, relaxing, not actually thinking, just coding.

**Comment by SethTisue** [comment]
@retronym fresh round of review?

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 4]
```diff
@@ -1276,8 +1276,8 @@ trait Infer extends Checkable {
      *  If no alternative matches `pt`, take the parameterless one anyway.
      */
     def inferExprAlternative(tree: Tree, pt: Type): Tree = {
-      val c = context
-      class InferTwice(pre: Type, alts: List[Symbol]) extends c.TryTwice {
+      val `context'` = context
+      class InferTwice(pre: Type, alts: List[Symbol]) extends `context'`.TryTwice {
```
Comment (soc) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: I think using

```
`context'`
```

as a name is quite non-standard ...

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Yes, I got carried away. But it's also annoying to fire up an IDE just to see where the letter `c` is used. Since it's just for the stable id alias, it's a use case for `prime'`. So, nothing progresses without taking some risks.

File: src/compiler/scala/tools/nsc/typechecker/Adaptations.scala [hunk 5]
```diff
@@ -26,20 +26,25 @@ trait Adaptations {
...
+          t.symbol.owner.decodedName
+        }${
+          if (t.symbol.isConstructor || t.symbol.name == nme.apply) "" else s".${ t.symbol.decodedName }"
+        }"
+      def sigString = s"${
+          t.symbol.owner.decodedName
+        }${
+          if (t.symbol.isConstructor) t.symbol.signatureString
+          else s".${t.symbol.decodedName}${t.symbol.signatureString}"
+        }"
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Adaptations.scala:47]: This is the other eccentric idiom. `}${` is the new `+`.

Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Adaptations.scala:47]: you mean the old `"+...+"` http://www.scala-lang.org/old/node/2025.html :)

**Comment by som-snytt** [comment]
@retronym Really no idea anymore how this compares to sliced bread.

**Comment by SethTisue** [comment]
some links with history on this: https://issues.scala-lang.org/browse/SI-2991, https://groups.google.com/forum/#!topic/scala-debate/zwG8o2YzCWs, https://github.com/scala/scala/pull/4232, https://github.com/scala/scala/issues/4242 ... surely an incomplete list...

**Comment by SethTisue** [comment]
I discussed with @retronym and I am sorry, but I think we won't be merging this, on the grounds that every such `-Y` option we add splinters Scala into dialects. and the alternate dialects don't have tooling support, or more than slight test coverage, and having them increases the overall compiler maintenance burden. so on the balance, it doesn't seem clearly superior to us to the existing workaround, which is to write a little shim in Java.

on the other hand, we want to know where this tupling stuff is headed in future Scala and Dotty versions, just leaving it alone isn't satisfactory. we'll discuss at next Scala team meeting that @odersky attends.

**Comment by som-snytt** [comment]
@SethTisue The assumption was that auto-tupling ought to be deprecated, so this was a migration path. It's not big bang for the complexity dollar, but does go to the newbie experience. I wonder if there's a meta-stackoverflow Q&A listing the Scala questions people are bored of answering. Maybe 2.13 is the version where anti-features can be dropped tout-court.

I think I mentioned this b/c @lrytz is in that area of the ballpark.

**Comment by adriaanm** [comment]
In light of Martin's comment at https://issues.scala-lang.org/browse/SI-2991, maybe we should reframe the discussion in terms of implementing a migration path to how dotty resolves overloading.

**Comment by som-snytt** [comment]
In retrospect, I see this as a tooling issue and not as a language issue. I'm returning to it because of a SO question which makes you realize that people rehearse these puzzlers on a daily basis, but they are invisible. Alternatively, one could describe it as a documentation issue. Or, as a tooling issue where a reporter is needed that detects the error and the correction, and then emits the required Java class. It would be neat if the reporter could retry the compilation.
